Startup fails due to IllegalArgumentException: Comparison method violates its general contract

**Jetty version(s)**
I have seen it 2 times out of a 1000 runs or so, once on jetty12ee9 beta3 and once on jetty12ee8 built today.
**Java version/vendor** `(use: java -version)`
JDK17
**OS type/version**
Unix
**Description**
Sometimes at startup server configuration throws an error. I think it might be some race condition? (It might also be caused by jrebel somehow).
Config error java.lang.IllegalArgumentException: Comparison method violates its general contract! at <Call class="org.eclipse.jetty.ee8.webapp.WebAppContext" name="addSystemClasses"><Arg><Ref refid="Environment"/></Arg><Arg>|      <Call class="org.eclipse.jetty.util.StringUtil" name="csvSplit"><Arg><Property name="jetty.webapp.addSystemClasses"/></Arg></Call>|    </Arg></Call> in file:///opt/containers/jetty-home-12.0.0.beta3/etc/jetty-ee8-webapp.xml

I will attach the console log file.

[jetty_console.log](https://github.com/eclipse/jetty.project/files/12160653/jetty_console.log)

```
2023-07-25 09:54:43.415:WARN :oejx.XmlConfiguration:main: Config error java.lang.IllegalArgumentException: Comparison method violates its general contract! at <Call class="org.eclipse.jetty.ee8.webapp.WebAppContext" name="addSystemClasses"><Arg><Ref refid="Environment"/></Arg><Arg>|      <Call class="org.eclipse.jetty.util.StringUtil" name="csvSplit"><Arg><Property name="jetty.webapp.addSystemClasses"/></Arg></Call>|    </Arg></Call> in file:///opt/containers/jetty-home-12.0.0.beta3/etc/jetty-ee8-webapp.xml
2023-07-25 09:54:43.420:WARN :oejx.XmlConfiguration:main: Unable to execute XmlConfiguration
java.lang.RuntimeException: java.lang.IllegalArgumentException: Comparison method violates its general contract!
	at org.eclipse.jetty.xml.XmlConfiguration.lambda$main$2(XmlConfiguration.java:1957)
	at org.eclipse.jetty.util.component.Environment.run(Environment.java:75)
	at org.eclipse.jetty.xml.XmlConfiguration.main(XmlConfiguration.java:1964)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.jetty.start.Main.invokeMain(Main.java:221)
	at org.eclipse.jetty.start.Main.start(Main.java:519)
	at org.eclipse.jetty.start.Main.main(Main.java:76)
Caused by: 
java.lang.IllegalArgumentException: Comparison method violates its general contract!
	at java.base/java.util.TimSort.mergeLo(TimSort.java:781)
	at java.base/java.util.TimSort.mergeAt(TimSort.java:518)
	at java.base/java.util.TimSort.mergeCollapse(TimSort.java:448)
	at java.base/java.util.TimSort.sort(TimSort.java:245)
	at java.base/java.util.Arrays.sort(Arrays.java:1233)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.call(XmlConfiguration.java:954)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.call(XmlConfiguration.java:932)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.configure(XmlConfiguration.java:490)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.configure(XmlConfiguration.java:443)
	at org.eclipse.jetty.xml.XmlConfiguration.configure(XmlConfiguration.java:345)
	at org.eclipse.jetty.xml.XmlConfiguration.lambda$main$2(XmlConfiguration.java:1950)
	at org.eclipse.jetty.util.component.Environment.run(Environment.java:75)
	at org.eclipse.jetty.xml.XmlConfiguration.main(XmlConfiguration.java:1964)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.jetty.start.Main.invokeMain(Main.java:221)
	at org.eclipse.jetty.start.Main.start(Main.java:519)
	at org.eclipse.jetty.start.Main.main(Main.java:76)
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.jetty.start.Main.invokeMain(Main.java:221)
	at org.eclipse.jetty.start.Main.start(Main.java:519)
	at org.eclipse.jetty.start.Main.main(Main.java:76)
Caused by: java.lang.RuntimeException: java.lang.IllegalArgumentException: Comparison method violates its general contract!
	at org.eclipse.jetty.xml.XmlConfiguration.lambda$main$2(XmlConfiguration.java:1957)
	at org.eclipse.jetty.util.component.Environment.run(Environment.java:75)
	at org.eclipse.jetty.xml.XmlConfiguration.main(XmlConfiguration.java:1964)
	... 7 more
Caused by: java.lang.IllegalArgumentException: Comparison method violates its general contract!
	at java.base/java.util.TimSort.mergeLo(TimSort.java:781)
	at java.base/java.util.TimSort.mergeAt(TimSort.java:518)
	at java.base/java.util.TimSort.mergeCollapse(TimSort.java:448)
	at java.base/java.util.TimSort.sort(TimSort.java:245)
	at java.base/java.util.Arrays.sort(Arrays.java:1233)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.call(XmlConfiguration.java:954)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.call(XmlConfiguration.java:932)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.configure(XmlConfiguration.java:490)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.configure(XmlConfiguration.java:443)
	at org.eclipse.jetty.xml.XmlConfiguration.configure(XmlConfiguration.java:345)
	at org.eclipse.jetty.xml.XmlConfiguration.lambda$main$2(XmlConfiguration.java:1950)
	... 9 more
java.lang.reflect.InvocationTargetException
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.eclipse.jetty.start.Main.invokeMain(Main.java:221)
	at org.eclipse.jetty.start.Main.start(Main.java:519)
	at org.eclipse.jetty.start.Main.main(Main.java:76)
Caused by: java.lang.RuntimeException: java.lang.IllegalArgumentException: Comparison method violates its general contract!
	at org.eclipse.jetty.xml.XmlConfiguration.lambda$main$2(XmlConfiguration.java:1957)
	at org.eclipse.jetty.util.component.Environment.run(Environment.java:75)
	at org.eclipse.jetty.xml.XmlConfiguration.main(XmlConfiguration.java:1964)
	... 7 more
Caused by: java.lang.IllegalArgumentException: Comparison method violates its general contract!
	at java.base/java.util.TimSort.mergeLo(TimSort.java:781)
	at java.base/java.util.TimSort.mergeAt(TimSort.java:518)
	at java.base/java.util.TimSort.mergeCollapse(TimSort.java:448)
	at java.base/java.util.TimSort.sort(TimSort.java:245)
	at java.base/java.util.Arrays.sort(Arrays.java:1233)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.call(XmlConfiguration.java:954)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.call(XmlConfiguration.java:932)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.configure(XmlConfiguration.java:490)
	at org.eclipse.jetty.xml.XmlConfiguration$JettyXmlConfiguration.configure(XmlConfiguration.java:443)
	at org.eclipse.jetty.xml.XmlConfiguration.configure(XmlConfiguration.java:345)
	at org.eclipse.jetty.xml.XmlConfiguration.lambda$main$2(XmlConfiguration.java:1950)
	... 9 more

Usage: java -jar $JETTY_HOME/start.jar [options] [properties] [configs]
       java -jar $JETTY_HOME/start.jar --help  # for more information
```

We start up jetty with a simple start.ini in the profile and it generates all other conf based on that.
If you feel like there is no race there, then you can probably close the case and it might be an issue caused by jrebels presence. But jrebels presence might make things slower and this can expose race conditions.

--- a/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java
+++ b/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java
@@ -119,31 +119,60 @@ public class XmlConfiguration
                     Class<?> t2 = p2[i].getType();
                     if (t1 != t2)
                     {
-                        // Compare distance from Object
-                        int d1 = distanceFromObject(t1);
-                        int d2 = distanceFromObject(t2);
-                        compare = Integer.compare(d2, d1);
+                        // prefer primitives
+                        compare = Boolean.compare(t2.isPrimitive(), t1.isPrimitive());
+                        if (compare == 0)
+                        {
+                            // prefer interfaces
+                            compare = Boolean.compare(t2.isInterface(), t1.isInterface());
 
-                        // break on the first different parameter (should always be true)
-                        if (compare != 0)
-                            break;
+                            if (compare == 0)
+                            {
+                                // prefer most derived
+                                int d1 = calculateDepth(t1);
+                                int d2 = calculateDepth(t2);
+                                compare = Integer.compare(d2, d1);
+                            }
+                        }
                     }
+                    // break on the first different parameter
+                    if (compare != 0)
+                        break;
                 }
             }
         }
+
+        // failsafe is to compare on the generic string
         if (compare == 0)
             compare = e1.toGenericString().compareTo(e2.toGenericString());
-        compare = Math.min(1, Math.max(compare, -1));
-        return compare;
+
+        // Return normalized to -1, 0, 1
+        return Integer.compare(compare, 0);
     };
 
-    private static int distanceFromObject(Class<?> c)
+    private static int calculateDepth(Class<?> c)
     {
         int depth = 0;
-        while (c != null && c != Object.class)
+
+        if (c.isPrimitive())
+            return Integer.MIN_VALUE;
+
+        if (c.isInterface())
         {
-            depth++;
-            c = c.getSuperclass();
+            Set<Class<?>> interfaces = Set.of(c);
+            while (!interfaces.isEmpty())
+            {
+                depth++;
+                interfaces = interfaces.stream().flatMap(i -> Arrays.stream(i.getInterfaces())).collect(Collectors.toSet());
+            }
+        }
+        else
+        {
+            while (c != Object.class && !c.isPrimitive())
+            {
+                depth++;
+                c = c.getSuperclass();
+            }
         }
         return depth;
     }
@@ -938,10 +967,11 @@ public class XmlConfiguration
                 throw new IllegalArgumentException("Method name cannot be blank");
 
             // Lets just try all methods for now
-            List<Method> methods = Arrays.stream(oClass.getMethods())
+            Method[] methods = Arrays.stream(oClass.getMethods())
                 .filter(m -> m.getName().equals(methodName))
                 .sorted(EXECUTABLE_COMPARATOR)
-                .toList();
+                .toArray(Method[]::new);
+
             for (Method method : methods)
             {
                 if (!method.getName().equals(methodName))

--- a/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java
+++ b/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java
@@ -119,16 +119,10 @@ public class XmlConfiguration
                     Class<?> t2 = p2[i].getType();
                     if (t1 != t2)
                     {
-                        // Favour derived type over base type
-                        compare = Boolean.compare(t1.isAssignableFrom(t2), t2.isAssignableFrom(t1));
-                        if (compare == 0)
-                        {
-                            // favour primitive type over reference
-                            compare = Boolean.compare(!t1.isPrimitive(), !t2.isPrimitive());
-                            if (compare == 0)
-                                // Use name to avoid non determinant sorting
-                                compare = t1.getName().compareTo(t2.getName());
-                        }
+                        // Compare distance from Object
+                        int d1 = distanceFromObject(t1);
+                        int d2 = distanceFromObject(t2);
+                        compare = Integer.compare(d2, d1);
 
                         // break on the first different parameter (should always be true)
                         if (compare != 0)
@@ -136,11 +130,23 @@ public class XmlConfiguration
                     }
                 }
             }
-            compare = Math.min(1, Math.max(compare, -1));
         }
-
+        if (compare == 0)
+            compare = e1.toGenericString().compareTo(e2.toGenericString());
+        compare = Math.min(1, Math.max(compare, -1));
         return compare;
     };
+
+    private static int distanceFromObject(Class<?> c)
+    {
+        int depth = 0;
+        while (c != null && c != Object.class)
+        {
+            depth++;
+            c = c.getSuperclass();
+        }
+        return depth;
+    }
 
     /**
      * Set the standard IDs and properties expected in a jetty XML file:

--- a/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java
+++ b/jetty-xml/src/main/java/org/eclipse/jetty/xml/XmlConfiguration.java
@@ -938,9 +938,10 @@ public class XmlConfiguration
                 throw new IllegalArgumentException("Method name cannot be blank");
 
             // Lets just try all methods for now
-
-            Method[] methods = oClass.getMethods();
-            Arrays.sort(methods, EXECUTABLE_COMPARATOR);
+            List<Method> methods = Arrays.stream(oClass.getMethods())
+                .filter(m -> m.getName().equals(methodName))
+                .sorted(EXECUTABLE_COMPARATOR)
+                .toList();
             for (Method method : methods)
             {
                 if (!method.getName().equals(methodName))

--- a/jetty-xml/src/test/java/org/eclipse/jetty/xml/XmlConfigurationTest.java
+++ b/jetty-xml/src/test/java/org/eclipse/jetty/xml/XmlConfigurationTest.java
@@ -729,6 +729,22 @@ public class XmlConfigurationTest
         {
         }
 
+        public void call(Aaa aaa)
+        {
+        }
+
+        public void call(Bbb aaa)
+        {
+        }
+
+        public void call(Ccc aaa)
+        {
+        }
+
+        public void call(Abc abc)
+        {
+        }
+
         public void call(Object o)
         {
         }
@@ -746,7 +762,7 @@ public class XmlConfigurationTest
         }
     }
 
-    @RepeatedTest(10)
+    @RepeatedTest(100)
     public void testMethodOrdering() throws Exception
     {
         List<Method> methods = Arrays.stream(TestOrder.class.getMethods()).filter(m -> "call".equals(m.getName())).collect(Collectors.toList());
@@ -755,7 +771,11 @@ public class XmlConfigurationTest
         assertThat(methods, Matchers.contains(
             TestOrder.class.getMethod("call"),
             TestOrder.class.getMethod("call", int.class),
+            TestOrder.class.getMethod("call", Abc.class),
+            TestOrder.class.getMethod("call", Aaa.class),
             TestOrder.class.getMethod("call", String.class),
+            TestOrder.class.getMethod("call", Bbb.class),
+            TestOrder.class.getMethod("call", Ccc.class),
             TestOrder.class.getMethod("call", Object.class),
             TestOrder.class.getMethod("call", String[].class),
             TestOrder.class.getMethod("call", String.class, String[].class)
@@ -1932,40 +1952,44 @@ public class XmlConfigurationTest
     @Test
     public void testExecutableComparator() throws Throwable
     {
-        xx(null);
-        yy(null);
-        zz(null);
+        aaa(null);
+        bbb(null);
+        ccc(null);
 
-        List<Method> methods = Arrays.stream(XmlConfigurationTest.class.getMethods())
-            .filter(m -> m.getName().length() == 2)
-            .toList();
+        List<Method> methods = Arrays.asList(Arrays.stream(XmlConfigurationTest.class.getMethods())
+            .filter(m -> m.getName().length() == 3)
+            .toArray(Method[]::new));
 
         // The implementor must also ensure that the relation is transitive: ((compare(x, y)>0) && (compare(y, z)>0)) implies compare(x, z)>0
         assertThat(EXECUTABLE_COMPARATOR.compare(methods.get(0), methods.get(1)), is(EXECUTABLE_COMPARATOR.compare(methods.get(1), methods.get(2))));
         assertThat(EXECUTABLE_COMPARATOR.compare(methods.get(0), methods.get(1)), is(EXECUTABLE_COMPARATOR.compare(methods.get(0), methods.get(2))));
     }
 
-    public void xx(XXX ignored)
+    public void aaa(Aaa ignored)
     {
     }
 
-    public void yy(YYY ignored)
+    public void bbb(Bbb ignored)
     {
     }
 
-    public void zz(ZZZ ignored)
+    public void ccc(Ccc ignored)
     {
     }
 
-    public interface XXX
+    public interface Aaa
     {
     }
 
-    public static class YYY
+    public interface Abc extends Aaa
     {
     }
 
-    public static class ZZZ implements XXX
+    public static class Bbb
+    {
+    }
+
+    public static class Ccc implements Aaa
     {
     }
 }

--- a/jetty-xml/src/test/java/org/eclipse/jetty/xml/XmlConfigurationTest.java
+++ b/jetty-xml/src/test/java/org/eclipse/jetty/xml/XmlConfigurationTest.java
@@ -59,6 +59,7 @@ import org.xml.sax.InputSource;
 import org.xml.sax.SAXException;
 
 import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.eclipse.jetty.xml.XmlConfiguration.EXECUTABLE_COMPARATOR;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.allOf;
 import static org.hamcrest.Matchers.containsString;
@@ -750,7 +751,7 @@ public class XmlConfigurationTest
     {
         List<Method> methods = Arrays.stream(TestOrder.class.getMethods()).filter(m -> "call".equals(m.getName())).collect(Collectors.toList());
         Collections.shuffle(methods);
-        methods.sort(XmlConfiguration.EXECUTABLE_COMPARATOR);
+        methods.sort(EXECUTABLE_COMPARATOR);
         assertThat(methods, Matchers.contains(
             TestOrder.class.getMethod("call"),
             TestOrder.class.getMethod("call", int.class),
@@ -1926,5 +1927,45 @@ public class XmlConfigurationTest
     private interface ThrowableAction
     {
         void run() throws Exception;
+    }
+
+    @Test
+    public void testExecutableComparator() throws Throwable
+    {
+        xx(null);
+        yy(null);
+        zz(null);
+
+        List<Method> methods = Arrays.stream(XmlConfigurationTest.class.getMethods())
+            .filter(m -> m.getName().length() == 2)
+            .toList();
+
+        // The implementor must also ensure that the relation is transitive: ((compare(x, y)>0) && (compare(y, z)>0)) implies compare(x, z)>0
+        assertThat(EXECUTABLE_COMPARATOR.compare(methods.get(0), methods.get(1)), is(EXECUTABLE_COMPARATOR.compare(methods.get(1), methods.get(2))));
+        assertThat(EXECUTABLE_COMPARATOR.compare(methods.get(0), methods.get(1)), is(EXECUTABLE_COMPARATOR.compare(methods.get(0), methods.get(2))));
+    }
+
+    public void xx(XXX ignored)
+    {
+    }
+
+    public void yy(YYY ignored)
+    {
+    }
+
+    public void zz(ZZZ ignored)
+    {
+    }
+
+    public interface XXX
+    {
+    }
+
+    public static class YYY
+    {
+    }
+
+    public static class ZZZ implements XXX
+    {
     }
 }

--- a/jetty-xml/src/test/java/org/eclipse/jetty/xml/XmlConfigurationTest.java
+++ b/jetty-xml/src/test/java/org/eclipse/jetty/xml/XmlConfigurationTest.java
@@ -17,17 +17,22 @@ import java.io.BufferedWriter;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
 import java.io.PrintStream;
+import java.lang.reflect.Executable;
 import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.net.URL;
 import java.nio.channels.SocketChannel;
+import java.nio.charset.Charset;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.time.Instant;
+import java.time.temporal.Temporal;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
 import java.util.HashMap;
 import java.util.List;
+import java.util.Locale;
 import java.util.Map;
 import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.ScheduledExecutorService;
@@ -62,6 +67,7 @@ import static java.nio.charset.StandardCharsets.UTF_8;
 import static org.eclipse.jetty.xml.XmlConfiguration.EXECUTABLE_COMPARATOR;
 import static org.hamcrest.MatcherAssert.assertThat;
 import static org.hamcrest.Matchers.allOf;
+import static org.hamcrest.Matchers.contains;
 import static org.hamcrest.Matchers.containsString;
 import static org.hamcrest.Matchers.empty;
 import static org.hamcrest.Matchers.instanceOf;
@@ -1956,6 +1962,13 @@ public class XmlConfigurationTest
         bbb(null);
         ccc(null);
 
+        Stream.of(XmlConfigurationTest.class.getMethods())
+            .filter(m -> m.getName().length() == 3)
+            .filter(m -> !m.getName().equals("foo"))
+            .sorted(EXECUTABLE_COMPARATOR)
+            .map(Executable::toGenericString)
+            .forEach(System.out::println);
+
         List<Method> methods = Arrays.asList(Arrays.stream(XmlConfigurationTest.class.getMethods())
             .filter(m -> m.getName().length() == 3)
             .toArray(Method[]::new));
@@ -1991,5 +2004,79 @@ public class XmlConfigurationTest
 
     public static class Ccc implements Aaa
     {
+    }
+
+    @Test
+    public void testFooExecutableComparator()
+    {
+        List<String> orderedMethodIds = Stream.of(FooObj.class.getMethods())
+            .filter(m -> m.getName().equals("foo"))
+            .sorted(EXECUTABLE_COMPARATOR)
+            .map(Executable::toGenericString)
+            .collect(Collectors.toList());
+        orderedMethodIds.forEach(System.out::println);
+        String[] expectedOrder = {
+            "public void org.eclipse.jetty.xml.XmlConfigurationTest$FooObj.foo()", // favor fewer args
+            "public void org.eclipse.jetty.xml.XmlConfigurationTest$FooObj.foo(int)", // favor primitives over non-primitives
+            "public void org.eclipse.jetty.xml.XmlConfigurationTest$FooObj.foo(java.time.temporal.Temporal)", // favor over Instant
+            "public int org.eclipse.jetty.xml.XmlConfigurationTest$FooObj.foo(java.lang.String)",
+            "public java.util.Locale org.eclipse.jetty.xml.XmlConfigurationTest$FooObj.foo(java.nio.charset.Charset)",
+            "public void org.eclipse.jetty.xml.XmlConfigurationTest$FooObj.foo(java.time.Instant)",
+            "public void org.eclipse.jetty.xml.XmlConfigurationTest$FooObj.foo(java.util.Locale)",
+            "public void org.eclipse.jetty.xml.XmlConfigurationTest$FooObj.foo(int,java.lang.String)",
+            "public void org.eclipse.jetty.xml.XmlConfigurationTest$FooObj.foo(java.lang.String,int)",
+            "public void org.eclipse.jetty.xml.XmlConfigurationTest$FooObj.foo(int,java.lang.String,java.lang.String)",
+            "public void org.eclipse.jetty.xml.XmlConfigurationTest$FooObj.foo(int,java.lang.String,java.lang.String,java.lang.Object)"
+        };
+        assertThat(orderedMethodIds, contains(expectedOrder));
+    }
+
+    public static class FooObj
+    {
+        public void foo()
+        {
+        }
+
+        public int foo(String name)
+        {
+            return -1;
+        }
+
+        public void foo(int id)
+        {
+        }
+
+        public void foo(Locale locale)
+        {
+        }
+
+        public void foo(Instant timestamp) // Instant extends from Temporal
+        {
+        }
+
+        public void foo(Temporal temporal)
+        {
+        }
+
+        public Locale foo(Charset charset)
+        {
+            return null;
+        }
+
+        public void foo(String name, int id)
+        {
+        }
+
+        public void foo(int id, String name)
+        {
+        }
+
+        public void foo(int id, String name, String description)
+        {
+        }
+
+        public void foo(int id, String name, String description, Object value)
+        {
+        }
     }
 }

